### PR TITLE
Addition operation for two `char`s to produce a `String`

### PR DESCRIPTION
**Pull request by Ruster-a11y**
This is a proposal to support the addition operation to be possible between two `char` variables ultimately producing a newly allocated `String`.

```rust
let a = '🎈';
let b = '🎉';
let c = a + b;

assert_eq!(c, "🎈🎉");
```

# Motivation
Ease of `char` manipulation to achieve `String` results.

### CODE REVIEW

File: src/liballoc/char.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/// Implements the `+` operator for concatenating two `char`s together.
+///
+/// This operation results into a new string being allocated with copies of the two `char`s in
+/// the respective order.
+///
+/// # Examples
+///
+/// ```
+/// let a = char::from("🎈");
+/// let b = char::from('🎉');
```
Comment (hellow554) [inline src/liballoc/char.rs]:
```
let a = '🎈';
let b = '🎉';
```

Comment (Ruster-a11y) [reply src/liballoc/char.rs]: Thanks @hellow554. Please check now.

Comment (BO41) [reply src/liballoc/char.rs]: it is better now, but you didn't fix what @hellow554 suggested. :wink: 
You should remove `char::from(`

Comment (Ruster-a11y) [reply src/liballoc/char.rs]: @BO41 Should that be done just in the comments part or everywhere else as well?

Comment (BO41) [reply src/liballoc/char.rs]: consistency is always good. So I assume everywhere

Comment (Ruster-a11y) [reply src/liballoc/char.rs]: Okay thanks @BO41 ! I did this for the tests as well. 😀

**Review by ghost — CHANGES_REQUESTED** [review]
you need to run tidy ($ `python x.py test src/tools/tidy`)
`tidy error: /checkout/src/liballoc/char.rs:23: trailing whitespace`

**Comment by hellow554** [comment]
There is currently one upvote and 4 downvotes (if you wan't to call them like that).

I'm also kind of against this, because it makes things implicit that should be explicit (creating a `String` is quiet expensive because of the heap allocation).
The next thing is, if you write something like `'a' + 'b' + 'c'` it won't work which is somewhat unexpected (but easy to understand).

If you need something like this for your project, why don't you use a macro instead (easiest solution that came to my mind):
```rust
macro_rules! create_string {
    ($($x:expr),*) => ([$($x),*].iter().collect::<String>())
}
```

**Comment by BO41** [comment]
@hellow554 I see your point. I expected there to be a downside and that's unfortunate.
Is there any chance to achieve something similar? Because I really like the idea of having such a simple way and for new people this could make it easier to get into Rust, since most other languages work that way :thinking:

**Comment by KodrAus** [comment]
Thanks for the PR @Ruster-a11y!

Our existing `+` and `+=` operator implementations for strings all use a type on the left-hand-side that could be a re-usable allocation. Personally, I'd be more on board with #66215, for pushing `char`s onto an existing string-like type, rather than creating one in an operator here.

**Comment by Ruster-a11y** [comment]
> The next thing is, if you write something like 'a' + 'b' + 'c' it won't work which is somewhat unexpected (but easy to understand).

@hellow554 Do you know about this? https://github.com/rust-lang/rust/pull/66215

In short, this is in conjunction with that other PR so that there is a shorthand for `char`s to be concatenated into a  `String` which can be further chained so that more `char`s can be added to the resulting `String`.

> I'm also kind of against this, because it makes things implicit that should be explicit (creating a String is quiet expensive because of the heap allocation).

Why wouldn't the developer *already* know producing a `String` has an heap allocation cost attached to it?

As per my point of view, giving the developer access to an easier baked-in approach for the manipulation of `char`s which ultimately produce `String`s is quite rational. Since `char`s are building blocks for any logical `string` -- it is an expected behavior for them to, again, ultimately produce `String`s.

As of now, the developer's code would look like this:

```rust
let my_string = String::new();
my_string,push(some_char);
my_string.push(another_char);
my_string.push(some_other_char);
// ...and so on
```

Whereas, this is more legible and the developer gets saved from having to deal with more code to read/write for a trivial thing:

```rust
let my_string = some_char + another_char + some_other_char;
```

@BO41 Thoughts?

**Comment by Ruster-a11y** [comment]
@hellow554 Also your suggested `iter().collect()` approach is *at least* 2x slower in the best case scenario vs. `String.push()` for `char`s.

The following screenshot shows `iter().collect()` taking ~55secs to compute final result compared to ~14secs computation time of `String.push()` for an iteration count of `1000000000`:
image

Source code for the test metric used: https://github.com/Ruster-a11y/rust-char-test

**Comment by hellow554** [comment]
> @hellow554 Do you know about this? #66215

Nope :) Thanks for the head up

> Why wouldn't the developer already know producing a String has an heap allocation cost attached to it?

That's not the point. The point is adding two types that aren't heap allocated will produce something that will be heap allocated by using the `+` operator. e.g. `3 + 2` doesn't make a heap allocation at all and you can chain them as much as you like.

> @hellow554 Also your suggested iter().collect() approach 

Yes and no. I suggested to use a macro approach. the iter collect thing is just the first thing that came to my mind. The implementation of that macro is highly up to you (or whoever creates it).
You can draw a similarity here with the `vec!` macro if you like.

**Comment by Ruster-a11y** [comment]
@hellow554 Thanks for the explanation. Much appreciated!

I suppose I would have to compile my own custom `rustc` variation to get these sort of functionalities working.

In the least there should've been some leeway left for `char` and `String` to be allowed to have developer implementations for overriding functionalities like the `+` and `+=` operator.

**Comment by JohnTitor** [comment]
Ping from triage: @KodrAus @Ruster-a11y any updates on this? Is it blocked by something?

File: src/liballoc/char.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+/// This operation results into a new `String` being allocated with copies of the two `char`s in
+/// the respective order.
+///
+/// # Examples
+///
+/// ```
+/// let a = '🎈';
+/// let b = '🎉';
+/// let c = a + b;
+/// // `c` is a newly allocated `String`
```
Comment (bjorn3) [inline src/liballoc/char.rs]:
```suggestion
/// assert_eq!(c, "🎈🎉");
```

Examples are automatically tested by rustdoc.

**Comment by ollie27** [comment]
Any `impl Add<char> for char` would have to be added in `core` because "only traits defined in the current crate can be implemented for arbitrary types" ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9892e0c1465769dd1f0c4d2cce2b64f9)). `String` isn't available in `core` so it's not possible to add the impl proposed here even if we wanted to.

**Comment by dtolnay** [comment]
The String + char PR ended up falling through, so I'll go ahead and close this one too because I believe we wouldn't want char + char without having String + char. The orphan rule makes this impl much more complicated anyway as @ollie27 mentioned, and I agree with several of the other objections in this thread. Thanks anyway @Ruster-a11y!

**Pull request closed without merging**